In [ ]:
# default_exp preprocessing.transform

# preprocessing.transform

In [ ]:
#hide
from IPython.core.display import display, HTML
display(HTML("<style>.container { width:85% !important; }</style>"))

In [ ]:
#hide
%reload_ext autoreload
%autoreload 2
%matplotlib inline

In [ ]:
#export
from lemonade.preprocessing.clean import *
from lemonade.preprocessing.vocab import *
from fastai.imports import *
from fastai import *
from datetime import date

In [ ]:
#hide
from nbdev.showdoc import *

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Import" data-toc-modified-id="Import-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Import</a></span></li><li><span><a href="#Load-Data" data-toc-modified-id="Load-Data-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Load Data</a></span><ul class="toc-item"><li><span><a href="#Load-Cleaned-Data" data-toc-modified-id="Load-Cleaned-Data-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Load Cleaned Data</a></span></li><li><span><a href="#Load-Vocabs" data-toc-modified-id="Load-Vocabs-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Load Vocabs</a></span></li></ul></li><li><span><a href="#Assemble-Single-Patient" data-toc-modified-id="Assemble-Single-Patient-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Assemble Single Patient</a></span><ul class="toc-item"><li><span><a href="#Collate-Codes-&amp;-Offsets" data-toc-modified-id="Collate-Codes-&amp;-Offsets-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Collate Codes &amp; Offsets</a></span><ul class="toc-item"><li><span><a href="#rec_dfs[[ptid]]-+-demographics[[ptid]]" data-toc-modified-id="rec_dfs[[ptid]]-+-demographics[[ptid]]-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span><code>rec_dfs[[ptid]]</code> + <code>demographics[[ptid]]</code></a></span><ul class="toc-item"><li><span><a href="#Single-Value-Returned---Check" data-toc-modified-id="Single-Value-Returned---Check-3.1.1.1"><span class="toc-item-num">3.1.1.1&nbsp;&nbsp;</span>Single Value Returned - Check</a></span></li></ul></li><li><span><a href="#collate_codes_offsts()" data-toc-modified-id="collate_codes_offsts()-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span><code>collate_codes_offsts()</code></a></span><ul class="toc-item"><li><span><a href="#Old---delete" data-toc-modified-id="Old---delete-3.1.2.1"><span class="toc-item-num">3.1.2.1&nbsp;&nbsp;</span>Old - delete</a></span></li><li><span><a href="#Tests" data-toc-modified-id="Tests-3.1.2.2"><span class="toc-item-num">3.1.2.2&nbsp;&nbsp;</span>Tests</a></span></li></ul></li><li><span><a href="#Numericalize-codes" data-toc-modified-id="Numericalize-codes-3.1.3"><span class="toc-item-num">3.1.3&nbsp;&nbsp;</span>Numericalize codes</a></span><ul class="toc-item"><li><span><a href="#Experiments" data-toc-modified-id="Experiments-3.1.3.1"><span class="toc-item-num">3.1.3.1&nbsp;&nbsp;</span>Experiments</a></span></li><li><span><a href="#get_codenums_offsts()" data-toc-modified-id="get_codenums_offsts()-3.1.3.2"><span class="toc-item-num">3.1.3.2&nbsp;&nbsp;</span><code>get_codenums_offsts()</code></a></span></li></ul></li><li><span><a href="#get_demographics()" data-toc-modified-id="get_demographics()-3.1.4"><span class="toc-item-num">3.1.4&nbsp;&nbsp;</span><code>get_demographics()</code></a></span></li><li><span><a href="#Tests" data-toc-modified-id="Tests-3.1.5"><span class="toc-item-num">3.1.5&nbsp;&nbsp;</span>Tests</a></span><ul class="toc-item"><li><span><a href="#Codes-&amp;-Offsets" data-toc-modified-id="Codes-&amp;-Offsets-3.1.5.1"><span class="toc-item-num">3.1.5.1&nbsp;&nbsp;</span>Codes &amp; Offsets</a></span></li><li><span><a href="#Numericalize-+-tensor" data-toc-modified-id="Numericalize-+-tensor-3.1.5.2"><span class="toc-item-num">3.1.5.2&nbsp;&nbsp;</span>Numericalize + tensor</a></span></li><li><span><a href="#Demographics" data-toc-modified-id="Demographics-3.1.5.3"><span class="toc-item-num">3.1.5.3&nbsp;&nbsp;</span>Demographics</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#ItemBase---Patient" data-toc-modified-id="ItemBase---Patient-4"><span class="toc-item-num">4&nbsp;&nbsp;</span><code>ItemBase</code> - Patient</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#.cuda(),-CUDA-Semantics,-Monkey-Patching-Lists-for-CUDA" data-toc-modified-id=".cuda(),-CUDA-Semantics,-Monkey-Patching-Lists-for-CUDA-4.0.1"><span class="toc-item-num">4.0.1&nbsp;&nbsp;</span><code>.cuda()</code>, CUDA Semantics, Monkey-Patching Lists for CUDA</a></span></li></ul></li><li><span><a href="#Class-Patient" data-toc-modified-id="Class-Patient-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Class <code>Patient</code></a></span><ul class="toc-item"><li><span><a href="#Tests" data-toc-modified-id="Tests-4.1.1"><span class="toc-item-num">4.1.1&nbsp;&nbsp;</span>Tests</a></span></li></ul></li></ul></li><li><span><a href="#ItemList---PatientList" data-toc-modified-id="ItemList---PatientList-5"><span class="toc-item-num">5&nbsp;&nbsp;</span><code>ItemList</code> - PatientList</a></span><ul class="toc-item"><li><span><a href="#Multiprocessing-Implementation" data-toc-modified-id="Multiprocessing-Implementation-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Multiprocessing Implementation</a></span><ul class="toc-item"><li><span><a href="#Chunking" data-toc-modified-id="Chunking-5.1.1"><span class="toc-item-num">5.1.1&nbsp;&nbsp;</span>Chunking</a></span></li><li><span><a href="#Multiprocessing-Logic" data-toc-modified-id="Multiprocessing-Logic-5.1.2"><span class="toc-item-num">5.1.2&nbsp;&nbsp;</span>Multiprocessing Logic</a></span><ul class="toc-item"><li><span><a href="#Useful-Links-+-Tips" data-toc-modified-id="Useful-Links-+-Tips-5.1.2.1"><span class="toc-item-num">5.1.2.1&nbsp;&nbsp;</span>Useful Links + Tips</a></span></li></ul></li></ul></li><li><span><a href="#Class-PatientList" data-toc-modified-id="Class-PatientList-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Class <code>PatientList</code></a></span><ul class="toc-item"><li><span><a href="#get_pckl_dir()-Util-Function" data-toc-modified-id="get_pckl_dir()-Util-Function-5.2.1"><span class="toc-item-num">5.2.1&nbsp;&nbsp;</span><code>get_pckl_dir()</code> Util Function</a></span></li><li><span><a href="#Awesome-&amp;-Fast-PatientList-class" data-toc-modified-id="Awesome-&amp;-Fast-PatientList-class-5.2.2"><span class="toc-item-num">5.2.2&nbsp;&nbsp;</span>Awesome &amp; Fast <code>PatientList</code> class</a></span><ul class="toc-item"><li><span><a href="#Test" data-toc-modified-id="Test-5.2.2.1"><span class="toc-item-num">5.2.2.1&nbsp;&nbsp;</span>Test</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#create_all_ptlists()" data-toc-modified-id="create_all_ptlists()-6"><span class="toc-item-num">6&nbsp;&nbsp;</span><code>create_all_ptlists()</code></a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Tests" data-toc-modified-id="Tests-6.0.1"><span class="toc-item-num">6.0.1&nbsp;&nbsp;</span>Tests</a></span><ul class="toc-item"><li><span><a href="#1K" data-toc-modified-id="1K-6.0.1.1"><span class="toc-item-num">6.0.1.1&nbsp;&nbsp;</span>1K</a></span></li><li><span><a href="#10K" data-toc-modified-id="10K-6.0.1.2"><span class="toc-item-num">6.0.1.2&nbsp;&nbsp;</span>10K</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Do-All-Preprocessing" data-toc-modified-id="Do-All-Preprocessing-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Do All Preprocessing</a></span></li><li><span><a href="#Export" data-toc-modified-id="Export-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Export</a></span></li></ul></div>

### Load Cleaned Data

In [ ]:
PATH_1K, PATH_10K

('./datasets/synthea/1K', './datasets/synthea/10K')

In [ ]:
CONDITIONS

{'diabetes': '44054006||START',
 'stroke': '230690007||START',
 'alzheimers': '26929004||START',
 'coronary_heart': '53741008||START'}

In [ ]:
train_dfs, valid_dfs, test_dfs = load_cleaned_ehrdata(PATH_1K)

In [ ]:
#[patients, patient_demographics, observations, allergies, careplans, medications, imaging_studies, procedures, conditions, immunizations]
all_dfs = train_dfs
patients_df, patient_demographics_df, all_rec_dfs = all_dfs[0], all_dfs[1], all_dfs[2:]

In [ ]:
# [display(df.head()) for df in all_dfs]

In [ ]:
patients_df.head()

patient   birthdate  diabetes_y  \
indx                                                                 
0     e4361ed5-5ef7-407d-b5d3-576155ed66cd  1988-05-16       False   
1     b28ca05e-a076-4868-a1af-168ae06b59ce  1990-06-12       False   
2     7c2b8c3b-fa87-44e9-b868-01db98dad56c  1988-09-05       False   
3     0731de61-1a8b-4135-83c0-2ace696bcac8  1934-11-27       False   
4     a119d309-9cac-415a-b9bd-c0cf10f15beb  1980-12-23       False   

      diabetes_age  stroke_y  stroke_age  alzheimers_y  alzheimers_age  \
indx                                                                     
0              NaN     False         NaN         False             NaN   
1              NaN     False         NaN         False             NaN   
2              NaN     False         NaN         False             NaN   
3              NaN     False         NaN         False             NaN   
4              NaN     False         NaN         False             NaN   

      coronary_heart_y  coronary_heart_age  
indx                                        
0                False                 NaN  
1                False                 NaN  
2                False                 NaN  
3                False                 NaN  
4                False                 NaN

In [ ]:
vals = all_dfs[0].iloc[1].values
vals

array(['b28ca05e-a076-4868-a1af-168ae06b59ce', '1990-06-12', False, nan,
       False, nan, False, nan, False, nan], dtype=object)

In [ ]:
vals[0], vals[1], vals[2], vals[4], vals[6], vals[8]

('b28ca05e-a076-4868-a1af-168ae06b59ce',
 '1990-06-12',
 False,
 False,
 False,
 False)

#### Load Vocabs

In [ ]:
vocab_list_1K = EhrVocabList.load(PATH_1K)
obs_vocab, alg_vocab, crpl_vocab, med_vocab, img_vocab, proc_vocab, cnd_vocab, imm_vocab = vocab_list_1K.records_vocabs
bday, bmonth, byear, marital, race, ethnicity, gender, birthplace, city, state, zipcode  = vocab_list_1K.demographics_vocabs
age_mean, age_std = vocab_list_1K.age_mean, vocab_list_1K.age_std

In [ ]:
#all emb dimensions to be passed into model to determine nh
demographics_dims, recs_dims, demographics_dims_width, recs_dims_width = get_all_emb_dims(EhrVocabList.load(PATH_1K))

In [ ]:
demographics_dims

[(33, 8),
 (14, 7),
 (124, 11),
 (5, 5),
 (7, 6),
 (25, 8),
 (4, 5),
 (205, 13),
 (211, 13),
 (3, 5),
 (200, 13)]

In [ ]:
recs_dims

[(550, 17),
 (27, 8),
 (54, 9),
 (224, 13),
 (11, 6),
 (128, 12),
 (201, 13),
 (20, 7)]

In [ ]:
demographics_dims_width, recs_dims_width

(94, 85)

## Assemble Single Patient

In [ ]:
# for t in data_tables: display(t.head())

In [ ]:
tst_ptid='0731de61-1a8b-4135-83c0-2ace696bcac8'

In [ ]:
tst_pt = patients_df[patients_df.patient == tst_ptid]
tst_pt

patient   birthdate  diabetes_y  \
indx                                                                 
3     0731de61-1a8b-4135-83c0-2ace696bcac8  1934-11-27       False   

      diabetes_age  stroke_y  stroke_age  alzheimers_y  alzheimers_age  \
indx                                                                     
3              NaN     False         NaN         False             NaN   

      coronary_heart_y  coronary_heart_age  
indx                                        
3                False                 NaN

In [ ]:
patient_demographics_df.loc[tst_ptid]

birthdate          1934-11-27
marital                     M
race                    white
ethnicity          portuguese
gender                      F
birthplace             Boston
city               Foxborough
state           Massachusetts
zip                      2035
age_now_days            30670
Name: 0731de61-1a8b-4135-83c0-2ace696bcac8, dtype: object

In [ ]:
tst_pt_birthdate = tst_pt.birthdate.values[0]
tst_pt_birthdate

'1934-11-27'

In [ ]:
tst_pt_diabetes, tst_pt_stroke, tst_pt_alzheimers, tst_pt_coronaryheart = \
    tst_pt.diabetes_y.values[0], tst_pt.stroke_y.values[0], tst_pt.alzheimers_y.values[0], tst_pt.coronary_heart_y.values[0]

In [ ]:
tst_pt_diabetes, tst_pt_stroke, tst_pt_alzheimers, tst_pt_coronaryheart

(False, False, False, False)

### Collate Codes & Offsets

#hide
### `rec_dfs[[ptid]]` + `demographics[[ptid]]`
**being done in `PatientList` - So this only for testing**

- **Filter out for this patient** 
 - df already filtered by cutoff age)
 - This is being done in `PatientList`, this is just for testing, **i.e. this will never be called**
- Use double `[[ptid]]` so that - even when single values are returned as a df

In [ ]:
#hide
def get_rec_dfs(all_rec_dfs, ptid):
    rec_dfs = []
    for rec_df in all_rec_dfs:
        try:
            rec_dfs.append(rec_df.loc[[ptid]])
        except KeyError:
            rec_dfs.append(pd.DataFrame())
    return rec_dfs

In [ ]:
%time rec_dfs = get_rec_dfs(all_rec_dfs, tst_ptid)

CPU times: user 32.4 ms, sys: 0 ns, total: 32.4 ms
Wall time: 32.2 ms


rec_dfs -- observations, allergies, careplans, medications, imaging_studies, procedures, conditions, immunizations

The following are empty for this ptid
- allergies - 1
- ~~imaging_studies - 4~~

In [ ]:
# for rec_df, name in zip(rec_dfs, FILENAMES[1:]):
#     print(f'{name}: {rec_df.shape}')
#     display(rec_df.tail())

In [ ]:
p_obs, p_alg, p_crpl, p_med, p_img, p_proc, p_cnd, p_immn = rec_dfs

In [ ]:
demograph_vector = patient_demographics_df.loc[tst_ptid]

In [ ]:
demograph_vector

birthdate          1934-11-27
marital                     M
race                    white
ethnicity          portuguese
gender                      F
birthplace             Boston
city               Foxborough
state           Massachusetts
zip                      2035
age_now_days            30670
Name: 0731de61-1a8b-4135-83c0-2ace696bcac8, dtype: object

In [ ]:
demograph_vector[6]

'Foxborough'

In [ ]:
demograph_vector.fillna('xxnone')

birthdate          1934-11-27
marital                     M
race                    white
ethnicity          portuguese
gender                      F
birthplace             Boston
city               Foxborough
state           Massachusetts
zip                      2035
age_now_days            30670
Name: 0731de61-1a8b-4135-83c0-2ace696bcac8, dtype: object

#### Single Value Returned - Check

In [ ]:
test_ptid2 = '7c2b8c3b-fa87-44e9-b868-01db98dad56c'

In [ ]:
test_ptid2_dfs = get_rec_dfs(all_rec_dfs, test_ptid2)

In [ ]:
for rec_df, name in zip(test_ptid2_dfs, FILENAMES[1:]):
    print(f'{name}: {rec_df.shape}')
    display(rec_df.tail())

observations: (283, 3)


code  \
patient                                                                     
7c2b8c3b-fa87-44e9-b868-01db98dad56c           21000-5||41.5||fL||numeric   
7c2b8c3b-fa87-44e9-b868-01db98dad56c       777-3||335.9||10*3/uL||numeric   
7c2b8c3b-fa87-44e9-b868-01db98dad56c          32207-3||488.3||fL||numeric   
7c2b8c3b-fa87-44e9-b868-01db98dad56c           32623-1||11.7||fL||numeric   
7c2b8c3b-fa87-44e9-b868-01db98dad56c  72166-2||Never smoker||xxxnan||text   

                                      age  age_months  
patient                                                
7c2b8c3b-fa87-44e9-b868-01db98dad56c   28         340  
7c2b8c3b-fa87-44e9-b868-01db98dad56c   28         340  
7c2b8c3b-fa87-44e9-b868-01db98dad56c   28         340  
7c2b8c3b-fa87-44e9-b868-01db98dad56c   28         340  
7c2b8c3b-fa87-44e9-b868-01db98dad56c   28         340

allergies: (0, 0)


Empty DataFrame
Columns: []
Index: []

careplans: (13, 3)


code  age  age_months
patient                                                              
7c2b8c3b-fa87-44e9-b868-01db98dad56c  53950000||STOP    8          98
7c2b8c3b-fa87-44e9-b868-01db98dad56c  91251008||STOP   11         141
7c2b8c3b-fa87-44e9-b868-01db98dad56c  91251008||STOP   14         173
7c2b8c3b-fa87-44e9-b868-01db98dad56c  53950000||STOP   19         232
7c2b8c3b-fa87-44e9-b868-01db98dad56c  91251008||STOP   25         306

medications: (21, 3)


code  age  age_months
patient                                                             
7c2b8c3b-fa87-44e9-b868-01db98dad56c   198405||STOP   11         141
7c2b8c3b-fa87-44e9-b868-01db98dad56c   313782||STOP   14         173
7c2b8c3b-fa87-44e9-b868-01db98dad56c  1043400||STOP   19         230
7c2b8c3b-fa87-44e9-b868-01db98dad56c   849574||STOP   25         306
7c2b8c3b-fa87-44e9-b868-01db98dad56c   861467||STOP   25         306

imaging_studies: (1, 3)


code  age  age_months
patient                                                        
7c2b8c3b-fa87-44e9-b868-01db98dad56c  72696002   25         305

procedures: (17, 3)


code  age  age_months
patient                                                               
7c2b8c3b-fa87-44e9-b868-01db98dad56c  428191000124101   13         159
7c2b8c3b-fa87-44e9-b868-01db98dad56c  428191000124101   14         171
7c2b8c3b-fa87-44e9-b868-01db98dad56c  428191000124101   22         268
7c2b8c3b-fa87-44e9-b868-01db98dad56c         74016001   25         305
7c2b8c3b-fa87-44e9-b868-01db98dad56c        699253003   25         305

conditions: (34, 3)


code  age  age_months
patient                                                               
7c2b8c3b-fa87-44e9-b868-01db98dad56c   10509002||STOP   19         230
7c2b8c3b-fa87-44e9-b868-01db98dad56c  195662009||STOP   19         235
7c2b8c3b-fa87-44e9-b868-01db98dad56c  444814009||STOP   21         254
7c2b8c3b-fa87-44e9-b868-01db98dad56c  444814009||STOP   22         266
7c2b8c3b-fa87-44e9-b868-01db98dad56c  444448004||STOP   25         306

immunizations: (40, 3)


code  age  age_months
patient                                                    
7c2b8c3b-fa87-44e9-b868-01db98dad56c   113   22         268
7c2b8c3b-fa87-44e9-b868-01db98dad56c    43   22         268
7c2b8c3b-fa87-44e9-b868-01db98dad56c   114   22         268
7c2b8c3b-fa87-44e9-b868-01db98dad56c   140   25         304
7c2b8c3b-fa87-44e9-b868-01db98dad56c   140   28         340

### `collate_codes_offsts()`

- For empty Code DFs like alg seen above - codes list will be 20 xxnones
- For non-empty Code DFs
 1. Collate codes by year
 2. Compute respective offsets

In [ ]:
#export
def collate_codes_offsts(rec_df, age_start, age_stop, age_in_months=False):
    """Return a single patient's EmbeddingBag lookup codes and offsets for the given age span and age units"""
    codes  = []
    offsts = [0]
    if rec_df.empty: 
        age_span = age_stop - age_start
        codes = ['xxnone'] * age_span
        offsts = list(range(age_span))
    else:
        for i in range(age_start, age_stop, 1):
            if age_in_months: res = (rec_df.code[rec_df.age_months == i]).values
            else            : res = (rec_df.code[rec_df.age == i]).values
            if len(res) > 0: 
                codes.extend(res)      
                if i < age_stop - 1: offsts.append(offsts[-1] + len(res))
            else:          
                codes.append('xxnone') 
                if i < age_stop - 1: offsts.append(offsts[-1] + 1)
    return codes, offsts

#### Old - delete

```python
#export
def collate_codes_offsts(rec_df, cutoff_age):
    codes  = []
    offsts = [0]
    if rec_df.empty: 
        codes = ['xxnone'] * cutoff_age
        offsts = list(range(cutoff_age))
    else:
        for i in range(cutoff_age):
            res = (rec_df.code[rec_df.age==i]).values
            if len(res)>0: 
                codes.extend(res)      # old --> for j in range(len(res)): codes.append(res[j])
                if i < cutoff_age - 1: offsts.append(offsts[-1]+len(res))
            else:          
                codes.append('xxnone') 
                if i < cutoff_age - 1: offsts.append(offsts[-1]+1)
    return codes, offsts
```

#### Tests

In [ ]:
%time all_codes_offsts = [collate_codes_offsts(df, age_start=0, age_stop=20) for df in rec_dfs]

CPU times: user 28.6 ms, sys: 3.4 ms, total: 32 ms
Wall time: 30.5 ms


In [ ]:
%time all_codes_offsts_1 = [collate_codes_offsts(df, age_start=0, age_stop=20) for df in [p_obs, p_alg, p_crpl, p_med, p_img, p_proc, p_cnd, p_immn]]

CPU times: user 28.4 ms, sys: 0 ns, total: 28.4 ms
Wall time: 27.7 ms


In [ ]:
assert all_codes_offsts == all_codes_offsts_1

In [ ]:
len(all_codes_offsts[0][0]), len(all_codes_offsts_1[0][0]), len(p_obs)

(236, 236, 2044)

In [ ]:
%time all_codes_offsts = [collate_codes_offsts(df, age_start=95, age_stop=105, age_in_months=True) for df in rec_dfs]

CPU times: user 13.9 ms, sys: 261 µs, total: 14.2 ms
Wall time: 13.9 ms


In [ ]:
all_codes_offsts[0][0], all_codes_offsts[0][1]

(['8302-2||136.6||cm||numeric',
  '72514-3||0.6||{score}||numeric',
  '29463-7||23.3||kg||numeric',
  '39156-5||12.5||kg/m2||numeric',
  '8462-4||82.4||mmHg||numeric',
  '8480-6||138.1||mmHg||numeric',
  '72166-2||Never smoker||xxxnan||text',
  'xxnone',
  'xxnone',
  'xxnone',
  '8302-2||138.6||cm||numeric',
  '72514-3||3.1||{score}||numeric',
  '29463-7||24.2||kg||numeric',
  '39156-5||12.6||kg/m2||numeric',
  '8462-4||70.2||mmHg||numeric',
  '8480-6||131.5||mmHg||numeric',
  '72166-2||Never smoker||xxxnan||text',
  'xxnone',
  'xxnone',
  'xxnone',
  'xxnone',
  'xxnone'],
 [0, 7, 8, 9, 10, 17, 18, 19, 20, 21])

### Numericalize codes
- Call to `vocab.numericalize()` must always be a list
- Guaranteed given - `codes` returned from `collate_codes_offsts()` is always a list even if a single item

**Do all** -- Given patient id

#### Experiments

In [ ]:
PATH_1K

'./datasets/synthea/1K'

In [ ]:
# p_obs.to_csv(f'{PATH_1K}/p_obs.csv')

In [ ]:
p_obs_collated = pd.DataFrame(all_codes_offsts[0][0])

In [ ]:
# p_obs_collated.to_csv(f'{PATH_1K}/p_obs_collated.csv')

In [ ]:
oc,oo = obs_vocab.numericalize(all_codes_offsts[0][0]), all_codes_offsts[0][1]

In [ ]:
p_obs_numericalized = pd.DataFrame(oc)

In [ ]:
# p_obs_numericalized.to_csv(f'{PATH_1K}/p_obs_numericalized.csv')

In [ ]:
# obs_vocab.vocab_df.to_csv(f'{PATH_1K}/obs_vocab_df.csv')

In [ ]:
len(p_obs), len(p_obs_collated), len(oc)

(2044, 236, 236)

In [ ]:
oc_1,oo_1 = obs_vocab.numericalize(all_codes_offsts_1[0][0]), all_codes_offsts_1[0][1]

In [ ]:
len(oc), len(oc_1), len(oo), len(oo_1)

(236, 236, 20, 20)

In [ ]:
# obs_vocab.textify(oc)
# oc

In [ ]:
len(oc),len(oo)

(236, 20)

In [ ]:
pc,po = proc_vocab.numericalize(all_codes_offsts[5][0]), all_codes_offsts[5][1]

In [ ]:
# pc

In [ ]:
len(pc),len(po)

(23, 20)

#### `get_codenums_offsts()`

In [ ]:
#export
def get_codenums_offsts(rec_dfs, all_vocabs, age_start, age_stop, age_in_months):

    all_codes_offsts = [collate_codes_offsts(df, age_start, age_stop, age_in_months) for df in rec_dfs]
    obs_vocab, alg_vocab, crpl_vocab, med_vocab, img_vocab, proc_vocab, cnd_vocab, imm_vocab = all_vocabs
    
    obs_c,  obs_o  = obs_vocab.numericalize (all_codes_offsts[0][0]), all_codes_offsts[0][1]
    alg_c,  alg_o  = alg_vocab.numericalize (all_codes_offsts[1][0]), all_codes_offsts[1][1]
    crpl_c, crpl_o = crpl_vocab.numericalize(all_codes_offsts[2][0]), all_codes_offsts[2][1]
    med_c,  med_o  = med_vocab.numericalize (all_codes_offsts[3][0]), all_codes_offsts[3][1]
    img_c,  img_o  = img_vocab.numericalize (all_codes_offsts[4][0]), all_codes_offsts[4][1]
    proc_c, proc_o = proc_vocab.numericalize(all_codes_offsts[5][0]), all_codes_offsts[5][1]
    cnd_c,  cnd_o  = cnd_vocab.numericalize (all_codes_offsts[6][0]), all_codes_offsts[6][1]
    imm_c,  imm_o  = imm_vocab.numericalize (all_codes_offsts[7][0]), all_codes_offsts[7][1]
    
    all_codenums = [obs_c,alg_c,crpl_c,med_c,img_c,proc_c,cnd_c,imm_c]
    all_offsts   = [obs_o,alg_o,crpl_o,med_o,img_o,proc_o,cnd_o,imm_o]
    
    return all_codenums, all_offsts

### `get_demographics()`

In [ ]:
#export
def get_demographics(demograph_vector, demographics_vocabs, age_mean, age_std):

    bday, bmonth, byear, marital, race, ethnicity, gender, birthplace, city, state, zipcode = demographics_vocabs
    demograph_vector = demograph_vector.fillna('xxnone')
    demographics = []
    
    birthdate = pd.Timestamp(demograph_vector[0])
    
    demographics.extend(bday.numericalize      ([birthdate.day]))
    demographics.extend(bmonth.numericalize    ([birthdate.month]))
    demographics.extend(byear.numericalize     ([birthdate.year]))
    demographics.extend(marital.numericalize   ([demograph_vector[1]]))
    demographics.extend(race.numericalize      ([demograph_vector[2]]))
    demographics.extend(ethnicity.numericalize ([demograph_vector[3]]))
    demographics.extend(gender.numericalize    ([demograph_vector[4]]))
    demographics.extend(birthplace.numericalize([demograph_vector[5]]))
    demographics.extend(city.numericalize      ([demograph_vector[6]]))
    demographics.extend(state.numericalize     ([demograph_vector[7]]))
    demographics.extend(zipcode.numericalize   ([demograph_vector[8]]))
    age = (demograph_vector[9] - age_mean) / age_std
    
    return demographics, age

### Tests

#### Codes & Offsets

In [ ]:
%time codenums, offsts = get_codenums_offsts(rec_dfs, vocab_list_1K.records_vocabs, age_start=0, age_stop=20, age_in_months=False)

CPU times: user 342 ms, sys: 0 ns, total: 342 ms
Wall time: 341 ms


In [ ]:
med_c, med_o = codenums[3],offsts[3]

In [ ]:
p_med_codes, p_med_offsts = collate_codes_offsts(p_med, age_start=0, age_stop=20)

In [ ]:
assert len(med_c) == len(p_med_codes)
assert med_o == p_med_offsts

In [ ]:
med_vocab.textify(med_c)[2]

('562251||START', {'Amoxicillin 250 MG / Clavulanate 125 MG Oral Tablet'})

In [ ]:
p_med_codes

['xxnone',
 'xxnone',
 '562251||START',
 '562251||STOP',
 'xxnone',
 'xxnone',
 'xxnone',
 'xxnone',
 'xxnone',
 '313782||START',
 '313782||STOP',
 'xxnone',
 'xxnone',
 'xxnone',
 'xxnone',
 'xxnone',
 '313782||START',
 '313782||STOP',
 'xxnone',
 '562251||START',
 '562251||STOP',
 'xxnone',
 'xxnone',
 'xxnone']

In [ ]:
p_med_offsts, med_o

([0, 1, 2, 4, 5, 6, 7, 8, 9, 11, 12, 13, 14, 15, 16, 18, 19, 21, 22, 23],
 [0, 1, 2, 4, 5, 6, 7, 8, 9, 11, 12, 13, 14, 15, 16, 18, 19, 21, 22, 23])

In [ ]:
len(p_med_codes),len(p_med_offsts)

(24, 20)

In [ ]:
p_alg_codes, p_alg_offsts = collate_codes_offsts(p_alg, age_start=0, age_stop=20)

In [ ]:
p_alg_codes

['xxnone',
 'xxnone',
 'xxnone',
 'xxnone',
 'xxnone',
 'xxnone',
 'xxnone',
 'xxnone',
 'xxnone',
 'xxnone',
 'xxnone',
 'xxnone',
 'xxnone',
 'xxnone',
 'xxnone',
 'xxnone',
 'xxnone',
 'xxnone',
 'xxnone',
 'xxnone']

In [ ]:
p_alg_offsts

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]

In [ ]:
obs_c, obs_o = codenums[0],offsts[0]

In [ ]:
p_obs_codes, p_obs_offsts = collate_codes_offsts(p_obs, age_start=0, age_stop=20)

In [ ]:
len(p_obs_codes), len(p_obs_offsts), len(obs_c), len(obs_o)

(236, 20, 236, 20)

In [ ]:
assert obs_o == p_obs_offsts

In [ ]:
%time codenums1, offsts1 = get_codenums_offsts(rec_dfs, vocab_list_1K.records_vocabs, age_start=20, age_stop=220, age_in_months=True)

CPU times: user 524 ms, sys: 0 ns, total: 524 ms
Wall time: 520 ms


In [ ]:
for codenum, offst in zip(codenums1, offsts1):
    print(len(codenum))
    assert len(offst) == 200

346
200
201
202
200
200
203
200


#### Numericalize + tensor

In [ ]:
p_med_indxs = med_vocab.numericalize(p_med_codes)
p_med_indxs

[0,
 0,
 13,
 141,
 0,
 0,
 0,
 0,
 0,
 4,
 133,
 0,
 0,
 0,
 0,
 0,
 4,
 133,
 0,
 13,
 141,
 0,
 0,
 0]

In [ ]:
med_c

[0,
 0,
 13,
 141,
 0,
 0,
 0,
 0,
 0,
 4,
 133,
 0,
 0,
 0,
 0,
 0,
 4,
 133,
 0,
 13,
 141,
 0,
 0,
 0]

In [ ]:
assert len(p_med_codes) == len(p_med_indxs)

In [ ]:
assert p_med_indxs == med_c

In [ ]:
torch.tensor(p_med_indxs)

tensor([  0,   0,  13, 141,   0,   0,   0,   0,   0,   4, 133,   0,   0,   0,
          0,   0,   4, 133,   0,  13, 141,   0,   0,   0])

#### Demographics

In [ ]:
demograph_vector

birthdate          1934-11-27
marital                     M
race                    white
ethnicity          portuguese
gender                      F
birthplace             Boston
city               Foxborough
state           Massachusetts
zip                      2035
age_now_days            30670
Name: 0731de61-1a8b-4135-83c0-2ace696bcac8, dtype: object

In [ ]:
get_demographics(demograph_vector, vocab_list_1K.demographics_vocabs, vocab_list_1K.age_mean, vocab_list_1K.age_std)

([28, 12, 36, 2, 2, 4, 3, 5, 4, 2, 4], 1.574773090930408)

In [ ]:
dem_vector = ['1988-05-16','S',None,None,'F','North Adams','Marlborough','Massachusetts',1901,11141]

In [ ]:
dem_vector = pd.Series(dem_vector)
dem_vector

0       1988-05-16
1                S
2             None
3             None
4                F
5      North Adams
6      Marlborough
7    Massachusetts
8             1901
9            11141
dtype: object

In [ ]:
get_demographics(dem_vector, vocab_list_1K.demographics_vocabs, vocab_list_1K.age_mean, vocab_list_1K.age_std)

([17, 6, 90, 3, 0, 0, 3, 90, 8, 2, 48], -0.5053754917098182)

In [ ]:
get_demographics(patient_demographics_df.loc['e8b8ebb1-a7b9-4ce8-8c6d-cfd74a284ebf'], vocab_list_1K.demographics_vocabs, \
                 vocab_list_1K.age_mean, vocab_list_1K.age_std) #has null

([20, 6, 112, 0, 2, 4, 3, 10, 9, 2, 9], -1.361550119558768)

In [ ]:
# vocab_list_1K.demographics_vocabs[10].ctoi

# `ItemBase` - Patient

**Questions / TODOs**
1. ~~Delete y from here and only do in labelling~~
2. ~~Delete patient id~~
3. ~~Either implement `cuda()` or add `cuda()` call to each in `__init__()`~~
 - See below for details
 - Implemented separate method, to be called in `fit()`
 - <span class="burk">Not sure if this is a good idea</span>
4. **Implement a good `__repr__()`**
5. Clean up more
 - Each of **obs, alg, etc could be positional** once everything is tested out and functional
 - Cleans up this class significantly

### `.cuda()`, CUDA Semantics, Monkey-Patching Lists for CUDA

**CUDA**
- [`torch.Tensor.cuda()`](https://pytorch.org/docs/stable/tensors.html#torch.Tensor.cuda)
- [`torch.nn.Module.cuda()`](https://pytorch.org/docs/stable/nn.html#torch.nn.Module.cuda)
- [CUDA Semantics](https://pytorch.org/docs/stable/notes/cuda.html#cuda-semantics)

*CUDA - summary*
- the `.cuda()` call puts tensor or module (model) onto GPU 
 - if not already there else returns existing copy from GPU
 
**Not Monkey-Patching Python List**
- Decided to drop the idea
- Not straightforward for Python list object - if possible at all
- No time to research that now
- And most importantly - **future versions will use fastai L or something built on that**

## Class `Patient`

In [ ]:
#export
class Patient():
    def __init__(self, nums, offsts, demographics, age_now, birthdate, diabetes, stroke, alzheimers, coronaryheart, ptid):
        self.obs_nums  = torch.tensor(nums[0])
        self.alg_nums  = torch.tensor(nums[1])
        self.crpl_nums = torch.tensor(nums[2])
        self.med_nums  = torch.tensor(nums[3])
        self.img_nums  = torch.tensor(nums[4])
        self.proc_nums = torch.tensor(nums[5])
        self.cnd_nums  = torch.tensor(nums[6])
        self.imm_nums  = torch.tensor(nums[7])

        self.obs_offsts  = torch.tensor(offsts[0])
        self.alg_offsts  = torch.tensor(offsts[1])
        self.crpl_offsts = torch.tensor(offsts[2])
        self.med_offsts  = torch.tensor(offsts[3])
        self.img_offsts  = torch.tensor(offsts[4])
        self.proc_offsts = torch.tensor(offsts[5])
        self.cnd_offsts  = torch.tensor(offsts[6])
        self.imm_offsts  = torch.tensor(offsts[7])
        
        self.demographics = torch.tensor(demographics)
        self.age_now      = torch.tensor([age_now])

        self.ptid = ptid
        self.birthdate = birthdate
        self.diabetes = diabetes
        self.stroke = stroke
        self.alzheimers = alzheimers
        self.coronaryheart = coronaryheart
    
    def __repr__(self):
        return f'ptid:{self.ptid}, birthdate:{self.birthdate}, diabetes:{self.diabetes}, device:{self.alg_nums.device}'
    
    @classmethod
    def create(cls, rec_dfs, demograph, vocablist, ptid, birthdate, diabetes, stroke, alzheimers, coronaryheart, age_start, age_stop, age_in_months):
        codenums, offsts  = get_codenums_offsts(rec_dfs, vocablist.records_vocabs, age_start, age_stop, age_in_months)
        demographics, age_now = get_demographics(demograph, vocablist.demographics_vocabs, vocablist.age_mean, vocablist.age_std)
        return cls(codenums, offsts, demographics, age_now, birthdate, diabetes, stroke, alzheimers, coronaryheart, ptid)
    
    def to_gpu(self):
        self.obs_nums  = self.obs_nums.cuda()
        self.alg_nums  = self.alg_nums.cuda()
        self.crpl_nums = self.crpl_nums.cuda()
        self.med_nums  = self.med_nums.cuda()
        self.img_nums  = self.img_nums.cuda()
        self.proc_nums = self.proc_nums.cuda()
        self.cnd_nums  = self.cnd_nums.cuda()
        self.imm_nums  = self.imm_nums.cuda()

        self.obs_offsts  = self.obs_offsts.cuda()
        self.alg_offsts  = self.alg_offsts.cuda()
        self.crpl_offsts = self.crpl_offsts.cuda()
        self.med_offsts  = self.med_offsts.cuda()
        self.img_offsts  = self.img_offsts.cuda()
        self.proc_offsts = self.proc_offsts.cuda()
        self.cnd_offsts  = self.cnd_offsts.cuda()
        self.imm_offsts  = self.imm_offsts.cuda()
        
        self.demographics = self.demographics.cuda()
        self.age_now      = self.age_now.cuda()
        
        return self

### Tests

In [ ]:
%time p1 = Patient.create(rec_dfs, demograph_vector, vocab_list_1K, tst_ptid, tst_pt_birthdate, tst_pt_diabetes, \
                                  tst_pt_stroke, tst_pt_alzheimers, tst_pt_coronaryheart, age_start=0, age_stop=20, age_in_months=False)

CPU times: user 336 ms, sys: 0 ns, total: 336 ms
Wall time: 334 ms


In [ ]:
len(p1.obs_nums)

236

In [ ]:
assert len(p1.obs_nums) == len(p_obs_codes)

In [ ]:
p1.obs_offsts

tensor([  0,  45,  63,  77,  91,  98, 116, 123, 130, 137, 144, 151, 169, 176,
        183, 190, 197, 215, 222, 229])

In [ ]:
p1.diabetes, p1.ptid, p1.birthdate

(False, '0731de61-1a8b-4135-83c0-2ace696bcac8', '1934-11-27')

In [ ]:
p1

ptid:0731de61-1a8b-4135-83c0-2ace696bcac8, birthdate:1934-11-27, diabetes:False, device:cpu

In [ ]:
p1.demographics

tensor([28, 12, 36,  2,  2,  4,  3,  5,  4,  2,  4])

In [ ]:
p1.age_now

tensor([1.5748], dtype=torch.float64)

In [ ]:
p1.demographics.shape, p1.age_now.shape

(torch.Size([11]), torch.Size([1]))

In [ ]:
torch.cat((p1.demographics, p1.age_now.type(torch.LongTensor)), dim=0)

tensor([28, 12, 36,  2,  2,  4,  3,  5,  4,  2,  4,  1])

In [ ]:
p1.to_gpu()

ptid:0731de61-1a8b-4135-83c0-2ace696bcac8, birthdate:1934-11-27, diabetes:False, device:cuda:0

In [ ]:
p1.diabetes, p1.ptid, p1.obs_offsts, p1.alg_nums

(False,
 '0731de61-1a8b-4135-83c0-2ace696bcac8',
 tensor([  0,  45,  63,  77,  91,  98, 116, 123, 130, 137, 144, 151, 169, 176,
         183, 190, 197, 215, 222, 229], device='cuda:0'),
 tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        device='cuda:0'))

In [ ]:
p1.demographics, p1.age_now

(tensor([28, 12, 36,  2,  2,  4,  3,  5,  4,  2,  4], device='cuda:0'),
 tensor([1.5748], device='cuda:0', dtype=torch.float64))

# `ItemList` - PatientList

**Fixes needed to de-crappify**
- `ItemList` needs inherit from `L` which uses [`listify()`](https://github.com/fastai/fastai/blob/b118a611729d32a45dd1293f416d6e743df109be/fastai/core.py#L129)
- `PatientList` needs to inherit `ItemList`
- Cleanup up `__repr__`, `__iter__`, etc

**Performance - DONE for now**
- Add multiprocessing to use multiple CPU cores
 - [Python Concurrency - Multiprocessing for CPU bound tasks](https://realpython.com/python-concurrency/#cpu-bound-multiprocessing-version)
 - [Using Multiprocessing to Make Python Code Faster](https://medium.com/@urban_institute/using-multiprocessing-to-make-python-code-faster-23ea5ef996ba)
 - 👍 - [Quick Tutorial: Python Multiprocessing](https://further-reading.net/2017/01/quick-tutorial-python-multiprocessing/)
 - [Ray library](https://towardsdatascience.com/10x-faster-parallel-python-without-python-multiprocessing-e5017c93cce1) based on Apache Arrow
- Upgrade pytorch - initial cuda call taking too long

**File descriptor sharing strategy issue - FIXED**
- [Solution](https://discuss.pytorch.org/t/runtimeerror-received-0-items-of-ancdata/4999/4)👍
 - `torch.multiprocessing.set_sharing_strategy('file_system')`
- [Another solution](https://github.com/pytorch/pytorch/issues/973#issuecomment-346405667) - increase `ulimit`
- [Yet another solution](https://github.com/pytorch/pytorch/issues/973#issuecomment-346439265) - Jeremy (Nov 22, 2017) - use `ThreadPool`
 - [fastai forum discussion about ThreadPool solution](https://github.com/fastai/fastai/issues/23#issuecomment-346405976)

## Multiprocessing Implementation

### Chunking

In [ ]:
less_patients = patients_df.iloc[:48]

In [ ]:
all_indxs = np.arange(len(less_patients))
cpu_cnt = 24
chnk_sz = int(len(all_indxs)/cpu_cnt) 
all_indxs, len(all_indxs), chnk_sz

(array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
        17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
        34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47]), 48, 2)

In [ ]:
indx_chunks = []
for i in range(0, len(all_indxs), chnk_sz):
    indx_chunks.append(list(all_indxs[i:i+chnk_sz]))

len(indx_chunks), len(indx_chunks[0])

(24, 2)

### Multiprocessing Logic

Current assumption of error is - main process dies and thus children see broken pipe error
- Let the parallelized sub proc in each core load all data and do the heavy lifting
- The main proc just sends a list of indxs (patients) to work on

In [ ]:
import torch.multiprocessing as multiprocessing
multiprocessing.set_sharing_strategy('file_system')
cpu_cnt = int(multiprocessing.cpu_count())

def subproc(indx_chnk):
    print(multiprocessing.current_process().name, 'start: ',indx_chnk[0], 'end: ',indx_chnk[-1], 'len: ',len(indx_chnk))
    return f'start:{indx_chnk[0]} -- len:{len(indx_chnk)}'

def driverproc(total_pts):
    indx_chnks = []
    all_indxs = np.arange(total_pts)
    chnk_sz = int(total_pts / cpu_cnt)
    for i in range(0, total_pts, chnk_sz):
        indx_chnks.append(list(all_indxs[i:i+chnk_sz]))
    pool = multiprocessing.Pool(processes=cpu_cnt)
    rs = pool.map(subproc, indx_chnks)
    pool.close()
    print(rs)

In [ ]:
driverproc(9969) #(len(data_tables[0]))

ForkPoolWorker-1 start:  0 end:  414 len:  415
ForkPoolWorker-2 start:  415 end:  829 len:  415
ForkPoolWorker-3 start:  830 end:  1244 len:  415
ForkPoolWorker-4 start:  1245 end:  1659 len:  415
ForkPoolWorker-5 start:  1660 end:  2074 len:  415
ForkPoolWorker-6 start:  2075 end:  2489 len:  415
ForkPoolWorker-7 start:  2490 end:  2904 len:  415
['start:0 -- len:415', 'start:415 -- len:415', 'start:830 -- len:415', 'start:1245 -- len:415', 'start:1660 -- len:415', 'start:2075 -- len:415', 'start:2490 -- len:415', 'start:2905 -- len:415', 'start:3320 -- len:415', 'start:3735 -- len:415', 'start:4150 -- len:415', 'start:4565 -- len:415', 'start:4980 -- len:415', 'start:5395 -- len:415', 'start:5810 -- len:415', 'start:6225 -- len:415', 'start:6640 -- len:415', 'start:7055 -- len:415', 'start:7470 -- len:415', 'start:7885 -- len:415', 'start:8300 -- len:415', 'start:8715 -- len:415', 'start:9130 -- len:415', 'start:9545 -- len:415', 'start:9960 -- len:9']
ForkPoolWorker-8 start:  2905 e

So for 1,108 pts
- 24 proc processed 46 pts each = 1,104
- then 1 proc (ForkPoolWorker-2 above) processed 4 more => 1108 total

And for 9,969 pts
- 24 processes 415 pts each = 9,960 
- then 1 proc (ForkPoolWorker-1) processed 9 more = total 9969

#### Useful Links + Tips

- [Things I Wish They Told Me About Multiprocessing in Python](https://www.cloudcity.io/blog/2019/02/27/things-i-wish-they-told-me-about-multiprocessing-in-python/)
- Log into `./logs` folder
```python
f = open(f'log/{multiprocessing.current_process().name}_log.txt','a+')
f.write(f'{ptid} -started')
f.write(" --completed\r\n")
```

## Class `PatientList`

### `get_pckl_dir()` Util Function

**No need for this anymore**
```python
#export
def list_files(directory, extension):
    return [f for f in os.listdir(directory) if f.endswith('.' + extension)]
```

In [ ]:
#export
def get_pckl_dir(path, split, age_start, age_stop, age_in_months):
    dir_name = ''
    dir_name += 'months' if age_in_months else 'years'
    dir_name += f'_{age_start}_to_{age_stop}'
    pckl_dir = Path(f'{path}/processed/{dir_name}/{split}')
    return pckl_dir

### Awesome & Fast `PatientList` class

In [ ]:
#export 
import torch.multiprocessing as multiprocessing
multiprocessing.set_sharing_strategy('file_system')
cpu_cnt = int(multiprocessing.cpu_count())

class PatientList():
    def __init__(self, pts, path, split, age_start, age_stop, age_in_months):
        self.items     = pts
        self.base_path = path
        self.split     = split 
        self.age_start = age_start
        self.age_stop  = age_stop
        self.age_type  = 'months' if age_in_months else 'years'

    def __len__(self): return len(self.items)
    def __iter__(self): return iter(self.items)
    def __getitem__(self, idx):
        if isinstance(idx, (int,slice)): return self.items[idx]
        if isinstance(idx[0],bool):
            assert len(idx)==len(self) # bool mask
            return [o for m,o in zip(idx,self.items) if m]
        return [self.items[i] for i in idx]
    def __repr__(self):
        res  = f'{self.__class__.__name__} ({len(self)} items)\n'
        res += f'base path:{self.base_path}; split:{self.split}; age span:{self.age_stop - self.age_start} {self.age_type}\n'
        res += f'age_start:{self.age_start}; age_stop:{self.age_stop}; age_type:{self.age_type}\n'
        for item in self.items[:10]:
            res += f'{item.__repr__()}\n'
        if len(self)>10: res = res[:-1]+ '...]'
        return res

    def _create_pts_chunk(indx_chnk, all_dfs, vocablist, pckl_dir, age_start, age_stop, age_in_months, verbose):

        pckl_f = open(f'{pckl_dir}/patients_{indx_chnk[0]}_{indx_chnk[-1]}.ptlist', 'wb')
        pts = []
         
        for indx in indx_chnk:
            vals = all_dfs[0].iloc[indx].values
            ptid, birthdate = vals[0], vals[1]
            diabetes, stroke, alzheimers, coronaryheart = vals[2], vals[4], vals[6], vals[8]
            
            rec_dfs = []
            for rec_df in all_dfs[2:]:
                try:
                    rec_dfs.append(rec_df.loc[[ptid]])
                except KeyError:
                    rec_dfs.append(pd.DataFrame())
            
            demograph = all_dfs[1].loc[ptid]
            pts.append(Patient.create(rec_dfs, demograph, vocablist, ptid, birthdate, diabetes, stroke, alzheimers, coronaryheart, age_start, age_stop, age_in_months))
        
        pickle.dump(pts,pckl_f)
        pckl_f.close()
        if verbose: print(f'{multiprocessing.current_process().name}-- completed {len(indx_chnk)} patients')
        return len(pts)

    @classmethod
    def create_save(cls, all_dfs, vocablist, pckl_dir, age_start, age_stop, age_in_months, verbose=False):
        
        pckl_dir.mkdir(parents=True, exist_ok=True)
        pts, indx_chnks = [], []
        
        total_pts = len(all_dfs[0])
        all_indxs = np.arange(total_pts)
        chnk_sz = total_pts // (cpu_cnt-1)
        for i in range(0, total_pts, chnk_sz):
            indx_chnks.append(list(all_indxs[i:i+chnk_sz]))
        
        pool = multiprocessing.Pool(processes=cpu_cnt)
        parallelize = partial(cls._create_pts_chunk, all_dfs=all_dfs, vocablist=vocablist, pckl_dir=pckl_dir, age_start=age_start, age_stop=age_stop, age_in_months=age_in_months, verbose=verbose)
        all_chunks = pool.map(parallelize, indx_chnks)
        pool.close()
        
        print(f'{sum(all_chunks)} total patients completed, saved patient list to {pckl_dir}')

    @classmethod
    def load(cls, path, split, age_start, age_stop, age_in_months):
        pckl_dir = get_pckl_dir(path, split, age_start, age_stop, age_in_months)
        ptlist = []
        for file in Path(pckl_dir).glob("*.ptlist"):
            with open(file, 'rb') as infile:
                ptlist.extend(pickle.load(infile))
                
        return(cls(ptlist, path, split, age_start, age_stop, age_in_months))

#### Test

In [ ]:
tst_pckl_dir = get_pckl_dir(PATH_1K, split='train', age_start=0, age_stop=20, age_in_months=False)

In [ ]:
tst_pckl_dir

PosixPath('datasets/synthea/1K/processed/years_0_to_20/train')

In [ ]:
%time PatientList.create_save(all_dfs, vocab_list_1K, tst_pckl_dir, age_start=0, age_stop=20, age_in_months=False)

664 total patients completed, saved patient list to datasets/synthea/1K/processed/years_0_to_20/train
CPU times: user 1.75 s, sys: 821 ms, total: 2.57 s
Wall time: 17.6 s


# `create_all_ptlists()`

In [ ]:
#export
def create_all_ptlists(path:Path, age_start:int, age_stop:int, age_in_months:bool, vocab_path:Path=None, verbose:bool=False, delete_existing:bool=True):
    '''Create Patient Lists for train, valid and test'''
    if vocab_path is None: vocab_path = path
    all_dfs_splits = load_cleaned_ehrdata(path) #train_dfs, valid_dfs, test_dfs
    splits = ['train', 'valid', 'test']
    vocablist = EhrVocabList.load(vocab_path)
    
    for all_dfs, split in zip(all_dfs_splits, splits):
        pckl_dir = get_pckl_dir(path, split, age_start, age_stop, age_in_months)
        if delete_existing:
            for file in Path(pckl_dir).glob("*.ptlist"):
                file.unlink()
        PatientList.create_save(all_dfs, vocablist, pckl_dir, age_start, age_stop, age_in_months, verbose)

### Tests

#### 1K

In [ ]:
%time create_all_ptlists(PATH_1K, age_start=0, age_stop=20, age_in_months=False, verbose=False)

664 total patients completed, saved patient list to datasets/synthea/1K/processed/years_0_to_20/train
222 total patients completed, saved patient list to datasets/synthea/1K/processed/years_0_to_20/valid
222 total patients completed, saved patient list to datasets/synthea/1K/processed/years_0_to_20/test
CPU times: user 3.66 s, sys: 1.98 s, total: 5.64 s
Wall time: 32.4 s


In [ ]:
%time create_all_ptlists(PATH_1K, age_start=0, age_stop=35, age_in_months=False, verbose=False)

664 total patients completed, saved patient list to datasets/synthea/1K/processed/years_0_to_35/train
222 total patients completed, saved patient list to datasets/synthea/1K/processed/years_0_to_35/valid
222 total patients completed, saved patient list to datasets/synthea/1K/processed/years_0_to_35/test
CPU times: user 3.66 s, sys: 1.93 s, total: 5.59 s
Wall time: 40.2 s


In [ ]:
%time create_all_ptlists(PATH_1K, age_start=120, age_stop=360, age_in_months=True) #10 to 30 yrs in mos (seq_len = 240)

664 total patients completed, saved patient list to datasets/synthea/1K/processed/months_120_to_360/train
222 total patients completed, saved patient list to datasets/synthea/1K/processed/months_120_to_360/valid
222 total patients completed, saved patient list to datasets/synthea/1K/processed/months_120_to_360/test
CPU times: user 3.84 s, sys: 1.98 s, total: 5.82 s
Wall time: 46.8 s


In [ ]:
%time create_all_ptlists(PATH_1K, age_start=240, age_stop=360, age_in_months=True) #20 to 30 yrs in mos (seq_len = 120)

664 total patients completed, saved patient list to datasets/synthea/1K/processed/months_240_to_360/train
222 total patients completed, saved patient list to datasets/synthea/1K/processed/months_240_to_360/valid
222 total patients completed, saved patient list to datasets/synthea/1K/processed/months_240_to_360/test
CPU times: user 3.81 s, sys: 1.91 s, total: 5.72 s
Wall time: 22.8 s


In [ ]:
ptlist_1K = PatientList.load(PATH_1K, 'train', age_start=120, age_stop=360, age_in_months=True)

In [ ]:
len(ptlist_1K)

664

In [ ]:
ptlist_1K

PatientList (664 items)
base path:./datasets/synthea/1K; split:train; age span:240 months
age_start:120; age_stop:360; age_type:months
ptid:12b42348-29bb-4cf4-b878-d419a4c8f067, birthdate:1951-09-04, diabetes:False, device:cpu
ptid:4c59df3b-042b-42c2-901a-3783d6d77919, birthdate:1979-09-18, diabetes:False, device:cpu
ptid:8f07e577-7ab4-4e89-ae98-6d394b3929a9, birthdate:1967-01-13, diabetes:False, device:cpu
ptid:653753d9-a52f-4ef3-b285-bdc3c671293c, birthdate:1971-12-26, diabetes:False, device:cpu
ptid:1b801001-0bf9-4ad9-b175-9de9bc2d905f, birthdate:1979-11-26, diabetes:False, device:cpu
ptid:da6203bc-bb24-4628-9aed-2b851571ee36, birthdate:1999-03-14, diabetes:False, device:cpu
ptid:b02cd0ff-e891-4caa-b69b-2da7cb267578, birthdate:1957-09-12, diabetes:True, device:cpu
ptid:49a859d3-318a-4212-864e-cf7d17f5032c, birthdate:2006-04-24, diabetes:False, device:cpu
ptid:cb9293c4-243f-427e-ba70-6834c5872be6, birthdate:1952-10-21, diabetes:True, device:cpu
ptid:cf82e51b-5c34-4010-b904-1583b009eb

In [ ]:
len(ptlist_1K.items[200].obs_nums), len(ptlist_1K.items[200].obs_offsts)

(326, 240)

In [ ]:
ptlist_1K.items[200].alg_nums

tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [ ]:
ptlist_1K[300].alg_nums

tensor([ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
         0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
         0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
         0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
         0,  0,  0,  0,  0,  0, 19, 22, 23,  0,  0,  0,  0,  0,  0,  0, 25,  0,
         0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
         0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
         0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
         0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
         0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
         0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
         0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
         0,  0,  0,  0,  0,  0,  0,  0, 

#### 10K

In [ ]:
%time create_all_ptlists(PATH_10K, age_start=0, age_stop=20, age_in_months=False, verbose=False)

6645 total patients completed, saved patient list to datasets/synthea/10K/processed/years_0_to_20/train
2215 total patients completed, saved patient list to datasets/synthea/10K/processed/years_0_to_20/valid
2216 total patients completed, saved patient list to datasets/synthea/10K/processed/years_0_to_20/test
CPU times: user 30.4 s, sys: 7.74 s, total: 38.1 s
Wall time: 5min 1s


In [ ]:
ptlist_10K_test = PatientList.load(PATH_10K, 'test', age_start=0, age_stop=20, age_in_months=False)

In [ ]:
ptlist_10K_test

PatientList (2216 items)
base path:./datasets/synthea/10K; split:test; age span:20 years
age_start:0; age_stop:20; age_type:years
ptid:34b27871-713c-41e2-816c-c74bdfad0e39, birthdate:1991-01-23, diabetes:False, device:cpu
ptid:d4bc2dec-8bf3-44ae-9c02-e3d27f917cbb, birthdate:2015-04-12, diabetes:False, device:cpu
ptid:3b91276a-5b1c-4e8f-98cc-ff5743c05f0d, birthdate:1989-10-14, diabetes:False, device:cpu
ptid:67672c74-d139-4b80-afa5-79866cc443b7, birthdate:1969-12-18, diabetes:False, device:cpu
ptid:81ab3bbb-5764-4d4d-9dbb-0c4ed1a49063, birthdate:2000-09-28, diabetes:False, device:cpu
ptid:898e007b-0c63-480e-9ce9-f64ed8ff0ae0, birthdate:1952-11-02, diabetes:False, device:cpu
ptid:f87dba81-400c-4959-90c0-bce963b24c95, birthdate:1975-08-02, diabetes:False, device:cpu
ptid:e8317720-ebd1-4bc1-9092-b9332361e8e3, birthdate:2008-09-06, diabetes:False, device:cpu
ptid:8439fc83-0086-4b86-9f68-871f0c789414, birthdate:1947-10-06, diabetes:False, device:cpu
ptid:efefc282-2add-405d-883b-ed83bdad0cc8,

In [ ]:
ptlist_10K_test[300].obs_nums

tensor([  2,   8,  12,  20,  24,  30,  34,  40,  45,  51,  54,  60,  65,  70,
        597,   2,   8,  12,  73,  78, 597,   2,   7,  12,  72,  77, 597,   3,
          8,  12,  73,  77, 597,   3,   7,  12,  73,  77, 597,   3,   8,  12,
         74,  78, 597,   3,   7,  12,  73,  77, 597,   3,   8,  12,  73,  78,
        597,   3,   7,  12,  73,  78, 597,   3,   7,  12,  82,  73,  78, 597,
          3,   8,  12,  82,  72,  77, 597,   3,   8,  12,  82,  72,  78, 597,
         88,   3,   8,  12,  82,  72,  78, 597,   4,   8,  12,  82,  73,  77,
        597,   4,   8,  12,  82,  73,  78,  21,  23,  30,  35,  37,  42,  50,
         54,  59,  63,  69, 597,   4,   7,  12,  82,  73,  78, 597,   4,   8,
         12,  82,  72,  78, 597,   0,   4,   7,  12,  82,  72,  78, 597,   4,
          8,  12,  82,  72,  78, 597,   4,   8,  12,  82,  72,  78, 597,   5,
          7,  12,  82,  73,  78,  18,  25,  29,  34,  38,  43,  50,  54,  58,
         65,  70, 597,   5,   9,  12,  82,  73,  77, 597,   5,  

# Do All Preprocessing

In [ ]:
#export
def preprocess_ehr_dataset(path, today, valid_pct=0.2, test_pct=0.2, obs_vocab_buckets=5,
                           age_start=0, age_stop=20, age_in_months=False, vocab_path=None, from_raw_data=False):
    if from_raw_data:
        print('------------------- Splitting and cleaning raw dataset -------------------')
        clean_raw_ehrdata(path, valid_pct, test_pct, today)
        print('------------------- Creating vocab lists -------------------')
        EhrVocabList.create(path, num_buckets=obs_vocab_buckets).save()
    else:
        print('Since data is pre-cleaned, skipping Cleaning, Splitting and Vocab-creation')
 
    print('------------------- Creating patient lists -------------------')
    create_all_ptlists(path, age_start, age_stop, age_in_months, vocab_path)

# Export

In [ ]:
#hide
from nbdev.export import *
notebook2script()

Converted 01_preprocessing_clean.ipynb.
Converted 02_preprocessing_vocab.ipynb.
Converted 03_preprocessing_transform.ipynb.
Converted index.ipynb.
